src/query/crypt_opfunc.c

```diff
@@ -697,3 +700,590 @@ crypt_generate_random_bytes (char *dest, int length)
+} evp_cipher = { {0, }, {0, } };  // Do not omit the this initialization settings.
+// *INDENT-ON*
+
+static int
+init_dblink_cipher (EVP_CIPHER_CTX ** ctx, const EVP_CIPHER ** cipher_type, bool is_aes_algorithm)
+{
+  static int is_init_done = 0;
+
+  if (is_init_done == 0)
+    {
+      memset (evp_cipher.nonce, 0x07, sizeof (evp_cipher.nonce));
+
+#if !defined(CS_MODE)
+      memcpy (evp_cipher.master_key, tde_Cipher.data_keys.log_key, TDE_MASTER_KEY_LENGTH);
+#else
+      memcpy (evp_cipher.master_key, dblink_Cipher_key.master_key, TDE_MASTER_KEY_LENGTH);
+#endif
+      is_init_done = 1;
+    }
+
+#if defined(CS_MODE)
+  if (dblink_Cipher_key.is_loaded == false)
+    {
+      int err;
+      extern int dblink_get_cipher_master_key ();	// declared in "network_interface_cl.c"
+      if ((err = dblink_get_cipher_master_key ()) != NO_ERROR)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, err, 0);
+	  return err;
+	}
+
+      memcpy (evp_cipher.master_key, dblink_Cipher_key.master_key, TDE_MASTER_KEY_LENGTH);
+    }
+#endif
+
+  if ((*ctx = EVP_CIPHER_CTX_new ()) == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_ENCRYPTION_LIB_FAILED, 1, crypt_lib_fail_info[CRYPT_LIB_INIT_ERR]);
+      return ER_ENCRYPTION_LIB_FAILED;
+    }
+
+  if (is_aes_algorithm)
+    {
+      *cipher_type = EVP_aes_256_ctr ();
+    }
+  else
+    {
+      *cipher_type = EVP_aria_256_ctr ();
+    }
+
+  if (*cipher_type == NULL)
+    {
+      EVP_CIPHER_CTX_free (*ctx);
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_ENCRYPTION_LIB_FAILED, 1, crypt_lib_fail_info[CRYPT_LIB_INIT_ERR]);
+      return ER_ENCRYPTION_LIB_FAILED;
+    }
+
+  return NO_ERROR;
+}
+
+/*
+ * crypt_dblink_encrypt ()  : Converts binary encrypted by AES/ARIA to readable string data. 
+ *
+ * return	      : NO_ERROR or error code.
+ * str(in)	      : password string data(In fact, the binary data).
+ * str_len(in)        : length of "src"
+ * cipher_buffer(out) : Encrypted binary data (by AES/ARIA)
+ * pk(in)             : The private key(master key) used to create the encrypted binary 
+ * 
+ * Remark:  
+ *          If pk is an empty string, the internal master key is used and the AES algorithm is used.
+ *          Otherwise, it is selected between AES and ARIA algorithms according to the value of a specific location.                             
+ */
+int
+crypt_dblink_encrypt (const unsigned char *str, int str_len, unsigned char *cipher_buffer, unsigned char *pk)
+{
+  int len, cipher_len, err;
+  EVP_CIPHER_CTX *ctx;
+  const EVP_CIPHER *cipher_type;
+  unsigned char master_key[TDE_MASTER_KEY_LENGTH] = { 0, };	// Do NOT omit this initialize.
+  unsigned char *key;
+
+  assert (pk);
+  err = init_dblink_cipher (&ctx, &cipher_type, (*pk) ? ((pk[13] & 0x01) == 0x00) : true);
+  if (err != NO_ERROR)
+    {
+      return err;
+    }
+  if (pk[0] == 0x00)
+    {
+      key = evp_cipher.master_key;
+    }
+  else
+    {
+      strcpy ((char *) master_key, (char *) pk);
```

Comment:
How about using strncpy().

Reply:
In the current code situation, the two codes below are the same, so there seems to be no meaning of change.
strncpy ((char *) master_key, (char *) pk, strlen(pk));
strcpy ((char *) master_key, (char *) pk);

The code below is a little different, in this case we need to ensure that the pk's garbage area is clean.
strncpy((char *) master_key, (char *) pk, DE_MASTER_KEY_LENGTH);

In order to keep even the garbage area clean, it is initialized with the code below, and after strcpy() is executed, the back part will be filled with nulls.
unsigned char master_key[TDE_MASTER_KEY_LENGTH] = { 0, };

Reply:
How about checking length of pk?

Reply:
Is it to be prepared in case the pk exceeds the buffer size when copying?

pk will come as a string of up to 20 bytes or 14 bytes.
It won't overflow the buffer, but I'll check it when I'm in debug mode.